### PR TITLE
Im2 12 home page

### PR DESCRIPTION
**Pull request by Bleak671**

### CODE REVIEW

File: React/my-react-app/src/components/Registration.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import { getHash } from '../Utils/getHash';
+import { useSelector, useDispatch } from 'react-redux';
+import { setState } from '../redux/AuthInfoReducer';
+import { handleChangeEmail, handleChangePassword, handleSubmitRegister } from '../Utils/formFunctions';
+
+export function Registration() {   
+  const dispatch = useDispatch();
+  const state = useSelector((state) => state.AuthInfo);
+  //render or redirect, depending on authorizing 
+  var token = sessionStorage.getItem("accessToken") == null
```
Comment (e44alex) [inline React/my-react-app/src/components/Registration.js:21]: Do not take token straight from storage. Bring in some isAuthenticated flag in state storage and check it every time you need

File: React/my-react-app/src/redux/AddPhotoInfoReducer.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import { createSlice } from "@reduxjs/toolkit";
+
+export const AddPhotoInfoReducer = createSlice({
+    name:"AddPhotoInfo",
+    initialState: {
```
Comment (e44alex) [inline React/my-react-app/src/redux/AddPhotoInfoReducer.js:5]: `    initialState: {
        value: {
            Link: null,
            Name: null
        }
    },
    reducers: {
        setState: (state, value) => {
            state.value = value.payload;
        }
    }`

this repeats at all reducers. should be extracted. DRY

File: React/my-react-app/src/components/TrashList.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+export function TrashList() {  
+  //constants
+  const dispatch = useDispatch();
+  const loading = useSelector((state) => state.TrashList);
+  const token = sessionStorage.getItem("accessToken").split("").map(shiftChar(-17)).join('');
+  const guid = sessionStorage.getItem("guid");
+
+  //load once after rendering
+  useEffect(() => { loadData(token, `https://localhost:44340/api/Trash/${guid}`, dispatch, setState) }, []);
```
Comment (e44alex) [inline React/my-react-app/src/components/TrashList.js:24]: keep https://localhost:44340/api as a global constant. On other deployment api address may not be the same

File: React/my-react-app/src/Utils/sort.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
+export function sortByName(loading, dispatch, setState) {
```
Comment (e44alex) [inline React/my-react-app/src/Utils/sort.js:1]: for all external functions: this should only do some logic, but not also dispatch and set state

File: React/my-react-app/src/components/Drawer.js [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+            </div>
+            <canvas id="canvas"> 
+
+            </canvas>
+        </div>
+        
+    );
+
+    //draw image
+    function draw(img){
```
Comment (e44alex) [inline React/my-react-app/src/components/Drawer.js:68]: extract this from component and create a function library

File: P4/P4Tests/PhotoBLLTests.cs [hunk 6]
```diff
@@ -23,7 +23,7 @@ public void Setup()
             .UseInMemoryDatabase(databaseName: Guid.NewGuid().ToString())
             .Options;
             _context = new PhotoRepository(new AppDBContext(options));
-            _context.Create(new Photo { PhotoId = Guid.Parse("31231234-1234-1242-1242-123412341234"), Link = "", UserId = Guid.Parse("11111111-1111-1111-1111-111111111111") });
+            _context.Create(new Photo { PhotoId = Guid.Parse("31231234-1234-1242-1242-123412341234"), Link = "", Hash="ascfdawdeqwda", UserId = Guid.Parse("11111111-1111-1111-1111-111111111111") });
```
Comment (e44alex) [inline P4/P4Tests/PhotoBLLTests.cs:26]: Need more test cases to check hash on same photos, etc.

File: React/my-react-app/src/App.js [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+import {AdminPhotoList} from './components/AdminPhotoList.js'
+import {AdminUserList} from './components/AdminUserList.js'
+import {AdminPhoto} from './components/AdminPhoto.js'
+import {AdminUser} from './components/AdminUser.js'
+import {Drawer} from './components/Drawer.js'
+import {NotificationContainer} from 'react-notifications';
+import "react-notifications/lib/notifications.css"
+
+function App() {
+  const GENRES = [
```
Comment (e44alex) [inline React/my-react-app/src/App.js:31]: Should extract routes

File: React/my-react-app/src/Utils/login.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  .then(res => res.json())
+  .then(
+    (response) => {
+        if (response.access_token != null) {
+          let base64ToString = Buffer.from(response.access_token.split('.')[1], "base64").toString();
+          base64ToString = JSON.parse(base64ToString);
+          sessionStorage.setItem("accessToken", response.access_token.split("").map(shiftChar(17)).join(''));
+          sessionStorage.setItem("guid", base64ToString["http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name"]);
+          sessionStorage.setItem("isAdmin", base64ToString["http://schemas.microsoft.com/ws/2008/06/identity/claims/role"]);
+          window.location.reload();
```
Comment (e44alex) [inline React/my-react-app/src/Utils/login.js:26]: there is an SPA, so no window.reload() allowed

File: React/my-react-app/src/Utils/login.js [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        }
+        else {
+          NotificationManager.error('Incorrect email/password', 'Log In',2000);
+        }     
+    }
+    ,(error) => {
+      NotificationManager.error('Fetch error', '',2000);
```
Comment (e44alex) [inline React/my-react-app/src/Utils/login.js:33]: should also be in saga

File: React/my-react-app/src/components/AdminPhoto.js [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+  const id = props.match.params.PhotoId.toString();
+  var options = {
+    year: 'numeric',
+    month: 'long',
+    day: 'numeric',
+    timezone: 'UTC',
+    hour: 'numeric',
+    minute: 'numeric',
+    second: 'numeric'
+  };
```
Comment (e44alex) [inline React/my-react-app/src/components/AdminPhoto.js:33]: conversion options should be used global and no be a part of component

File: React/my-react-app/src/App.js [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+      name: 'Sign in',
+    },
+    {
+      route: '/reg',
+      name: 'Sign up',
+    }
+  ]
+
+  const token = sessionStorage.getItem("accessToken");
+  const isAdmin = sessionStorage.getItem("isAdmin");
```
Comment (e44alex) [inline React/my-react-app/src/App.js:47]: should no be kept in session storage. keep in store

File: P4/P4Tests/PhotoBLLTests.cs (hunk 6 above)
Comment (Bleak671) [reply P4/P4Tests/PhotoBLLTests.cs:26]: hash is calculated on front, so I can't test it there without major changes

**Pull request merged**